### PR TITLE
Wallet: Activity Items - Sections

### LINKED ISSUE #19852: Wallet: Activity - Group into sections
The wallet activity items needs to be grouped into sections by date.

Designs: https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=1070-284225&mode=design&t=05G37UXdKebXGVOI-4

### PR DESCRIPTION
**Pull request by OmarBasem**
fixes: https://github.com/status-im/status-mobile/issues/19852

epic: https://github.com/status-im/status-mobile/issues/19849

### Summary
This PR implements sectioning for the activity items. 

#### Note: 
The other two issues in this epic will be out of scope from the July release.

[Designs](https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&mode=design&t=wQ9KoWf3M0AgvKNc-0)

### Demo:

https://github.com/status-im/status-mobile/assets/29354102/f17849c1-6e55-4237-9038-3b2547b1cba1

### CODE REVIEW

File: src/status_im/subs/wallet/activities_test.cljs
Comment (J-Son89) [inline src/status_im/subs/wallet/activities_test.cljs:10]: 🙌🏼

File: src/status_im/subs/wallet/activities.cljs
Comment (J-Son89) [inline src/status_im/subs/wallet/activities.cljs]: What is acts? It's better to just use full variable names 👍

File: src/status_im/subs/wallet/activities.cljs
Comment (ilmotta) [inline src/status_im/subs/wallet/activities.cljs]: No strong suggestion to change, but I felt like sharing the following tip for Clojure.

The code is a bit anticlimactic for Clojure because it's storing individual steps as separate bindings and not using the thread last macro, which was created exactly for this sort of scenario. Clojure devs are used to seeing data transformations as a linear sequence of steps. When we break the transformations into separate bindings we now force the dev to read the transformations in zig-zag fashion: filter ->  binding -> group -> binding -> map.

I suppose this was done to help with debugging because then the dev can inspect each binding individually. But when using the thread last macro we can also easily inspect individual steps without having to unbundle the whole pipeline into separate local bindings.

```clojure
(rf/reg-sub :wallet/activities-for-current-viewing-account
 :<- [:wallet/all-activities]
 :<- [:wallet/current-viewing-account-address]
 (fn [[activities current-viewing-account-address]]
   (->> activities
        (filter (fn [{:keys [sender recipient]}]
                  (or (= sender current-viewing-account-address)
                      (= recipient current-viewing-account-address))))
        user/inspect ; ***************** HERE Add this function between steps
        (group-by (fn [{:keys [timestamp]}]
                    (datetime/timestamp->relative-short-date (* timestamp 1000))))
        (map (fn [[date acts]]
               {:title date :data acts})))))

;; In src/user.cljs, the inspect function is defined:
(defn inspect
  [v]
  (tap> v)
  v)
```

File: src/status_im/subs/wallet/activities_test.cljs
Comment (ilmotta) [inline src/status_im/subs/wallet/activities_test.cljs]: The textual description should be on the same line as `testing`. There are other instances of this styling issue in this file.

File: src/status_im/subs/wallet/activities_test.cljs
Comment (ilmotta) [inline src/status_im/subs/wallet/activities_test.cljs]: Storing `result` seems overly verbose if it's going to be used only once and on the next line. The vast majority of other sub tests don't do this.

File: src/status_im/subs/wallet/activities_test.cljs
Comment (ilmotta) [inline src/status_im/subs/wallet/activities_test.cljs]: The value of `rf/sub` is already a deref'ed reaction. This test is throwing an exception because of this.

File: src/status_im/subs/wallet/activities_test.cljs
Comment (ilmotta) [inline src/status_im/subs/wallet/activities_test.cljs]: A few extra lines.

File: src/utils/datetime.cljs
Comment (ilmotta) [inline src/utils/datetime.cljs:212]: Please, cover utility functions with tests. I'd use the `are` macro to keep tests concise and assert lots of inputs.

File: src/status_im/subs/wallet/activities_test.cljs
Comment (ilmotta) [inline src/status_im/subs/wallet/activities_test.cljs]: I think this `assoc` is incorrect because the app-db paths are `[:wallet :activities]` and `[:wallet :current-viewing-account-address]`:

File: src/status_im/subs/wallet/activities.cljs
Comment (OmarBasem) [reply src/status_im/subs/wallet/activities.cljs]: Thanks for your review @ilmotta, been a long time :)

> The code is a bit anticlimactic for Clojure because it's storing individual steps as separate bindings

You are right, thanks for the suggestion, I will update it 👍 

> ;; In src/user.cljs, the inspect function is defined:

I am not sure where this function is? I don't see this function in the codebase. Or do you mean I need to add it?

Comment (ilmotta) [reply src/status_im/subs/wallet/activities.cljs]: Long long time, thanks for tagging me as a reviewer @OmarBasem.

The `user.cljs` is a common file Clojure devs use to create code to support their development, especially in REPLs, but it's not limited to REPL-Driven development. So you are correct, this file would need to be created and managed by you and you'd need to require this namespace somewhere, otherwise after every hot-reload the namespace definitions would be lost. Some teams even commit this file, but this is less common. It's also common for REPLs to start in the `user` namespace given their ubiquity. I assumed all of this was common knowledge, sorry for not explaining before.

Comment (OmarBasem) [reply src/status_im/subs/wallet/activities.cljs]: Understood @ilmotta! Thanks for explaining, and I see you made a PR for it already https://github.com/status-im/status-mobile/pull/19927 🙏

Comment (OmarBasem) [reply src/status_im/subs/wallet/activities.cljs]: Fixed!

File: src/utils/datetime_test.cljs
Comment (ilmotta) [inline src/utils/datetime_test.cljs:205]: Quick one: extra line

File: src/status_im/contexts/wallet/account/view.cljs
Comment (ilmotta) [inline src/status_im/contexts/wallet/account/view.cljs]: Should we delete the flag now that we don't have usages?

**Review by ilmotta — APPROVED** [review]
Code-wise looks good! Just one comment about deleting the feature flag. Thank you @OmarBasem

**Comment by pavloburykh** [comment]
@OmarBasem thanks for the PR. Please, take a look at the issues. 

### ISSUE 1 Only part of transactions are fetched in Activity tab

I have checked my old account that has a long history of activity on Mainnet. 

**Expected result:** 

According to Etherscan 37 transactions have been performed on Mainnet https://etherscan.io/txs?a=0x527d22094166ad33e7523d2cbca287798e149c7f

**Actual result:** only 18 transactions have been fetched in Activity tab in Status Mobile app. At the same time, all 37 transactions are fetched on latest Desktop master app. 

[Status-debug-logs - 2024-05-09T173052.604.zip](https://github.com/status-im/status-mobile/files/15263570/Status-debug-logs.-.2024-05-09T173052.604.zip)

https://github.com/status-im/status-mobile/assets/97245802/937834b8-121c-4be4-b5e4-bef803a73323

**Comment by pavloburykh** [comment]
### ISSUE 2 Some of transactions have wrong sorting by date

**Expected result:** transactions should be sorted by date in desc order (from latest to oldest) 

**Actual result:** on screenshot below you can see Oct 25, 2023 transaction appearing between Oct 6, 2022 and Sep 27, 2022 transactions. 

photo_2024-05-09 17 41 07

**Comment by J-Son89** [comment]
> @OmarBasem thanks for the PR. Please, take a look at the issues.
> 
> ### ISSUE 1 Only part of transactions are fetched in Activity tab
> I have checked my old account that has a long history of activity on Mainnet.
> 
> **Expected result:**
> 
> According to Etherscan 37 transactions have been performed on Mainnet https://etherscan.io/txs?a=0x527d22094166ad33e7523d2cbca287798e149c7f
> 
> **Actual result:** only 18 transactions have been fetched in Activity tab in Status Mobile app. At the same time, all 37 transactions are fetched on latest Desktop master app.
> 
> [Status-debug-logs - 2024-05-09T173052.604.zip](https://github.com/status-im/status-mobile/files/15263570/Status-debug-logs.-.2024-05-09T173052.604.zip)
> 
>  telegram-cloud-document-2-5326063766947777498.mp4

yeah I think we need to align with Desktop team on this. Pretty sure @smohamedjavid & I discussed this with @stefandunca and we will need to adjust a bit to get the rest of the activities that we are not seeing here.
In any case, I think the steps this pr takes is good for now and those missing activities can be considered beyond the scope of this pr and we can work with @smohamedjavid next week when he returns to gather all of the missing transaction data.

It's also clear we're missing some address details, i.e right now it just shows the address but it can be an account name or a saved address or a multichain address so we need to eventually handle those cases as well.

**Comment by pavloburykh** [comment]
> yeah I think we need to align with Desktop team on this. Pretty sure @smohamedjavid & I discussed this with @stefandunca and we will need to adjust a bit to get the rest of the activities that we are not seeing here. In any case, I think the steps this pr takes is good for now and those missing activities can be considered beyond the scope of this pr and we can work with @smohamedjavid next week when he returns to gather all of the missing transaction data.
> 
> It's also clear we're missing some address details, i.e right now it just shows the address but it can be an account name or a saved address or a multichain address so we need to eventually handle those cases as well.

@J-Son89 thanks! Are we tracking the issue with missing activities? Should it be the part of https://github.com/status-im/status-mobile/issues/19849 epic? 

Regarding this PR: according to @OmarBasem ```The other two issues in this epic will be out of scope from the July release.``` So this PR is the last one in epic https://github.com/status-im/status-mobile/issues/19849. That's why I assumed I need to test the entire Activity tab feature, as after merging this PR we are closing the epic. Am I right?

**Comment by J-Son89** [comment]
> > yeah I think we need to align with Desktop team on this. Pretty sure @smohamedjavid & I discussed this with @stefandunca and we will need to adjust a bit to get the rest of the activities that we are not seeing here. In any case, I think the steps this pr takes is good for now and those missing activities can be considered beyond the scope of this pr and we can work with @smohamedjavid next week when he returns to gather all of the missing transaction data.
> > It's also clear we're missing some address details, i.e right now it just shows the address but it can be an account name or a saved address or a multichain address so we need to eventually handle those cases as well.
> 
> @J-Son89 thanks! Are we tracking the issue with missing activities? Should it be the part of #19849 epic?
> 
> Regarding this PR: according to @OmarBasem `The other two issues in this epic will be out of scope from the July release.` So this PR is the last one in epic #19849. That's why I assumed I need to test the entire Activity tab feature, as after merging this PR we are closing the epic. Am I right?

good point @pavloburykh - apologies that's my fault. 🙏  I somewhat forgot about these missing transactions until you reminded me here! However I do recall these details were discussed now.

Let's add this issue you found as part of the same epic and we can close it then.
@OmarBasem is correct in that the other two issues are out of scope, I will adjust the epic to reflect that and pull them out to some separate work in the future. Apologies again, I should have  organised that epic in advance too, it's slightly confusing the way I organised it at this point. 👍

**Comment by pavloburykh** [comment]
> good point @pavloburykh - apologies that's my fault. 🙏

@J-Son89 no worries at all, not your fault. I have logged the issue https://github.com/status-im/status-mobile/issues/19956 and added it to https://github.com/status-im/status-mobile/issues/19852 epic.

**Comment by pavloburykh** [comment]
Meanwhile, @OmarBasem could you please take a look at [ISSUE 2](https://github.com/status-im/status-mobile/pull/19906#issuecomment-2102799878). It might be PR related. Thank you!

**Comment by OmarBasem** [comment]
> Meanwhile, @OmarBasem could you please take a look at [ISSUE 2](https://github.com/status-im/status-mobile/pull/19906#issuecomment-2102799878). It might be PR related. Thank you!

Hi @pavloburykh, thanks for your testing! 

Issue is fixed now!

**Comment by pavloburykh** [comment]
> > Meanwhile, @OmarBasem could you please take a look at [ISSUE 2](https://github.com/status-im/status-mobile/pull/19906#issuecomment-2102799878). It might be PR related. Thank you!
> 
> Hi @pavloburykh, thanks for your testing!
> 
> Issue is fixed now!

Thank you @OmarBasem! I will continue testing once we update go version in order to get the fix of updating balances on testnet. Currently balances are not updating in develop so I cannot test new transactions to make sure they appear in Activity tab.

**Comment by pavloburykh** [comment]
Hey @OmarBasem! Could you please rebase the branch? There are conflicts so I am unable to do it myself. Thank you.

**Comment by OmarBasem** [comment]
> Hey @OmarBasem! Could you please rebase the branch? There are conflicts so I am unable to do it myself. Thank you.

@pavloburykh Done!

**Comment by OmarBasem** [comment]
> > Hey @OmarBasem! Could you please rebase the branch? There are conflicts so I am unable to do it myself. Thank you.
> 
> @pavloburykh Done!

**Comment by pavloburykh** [comment]
> @pavloburykh Done!

Thanks @OmarBasem! Could you please take a look at failed integration tests https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-19906/9/consoleText ?

**Comment by pavloburykh** [comment]
Hey @OmarBasem! 

I am wondering if Activity items are ready for testing from design perspective? I see some differences compared to the designs. For example:

### ISSUE 3 Activity item misses circle
### ISSUE 4 Transaction status icon differs from the designs

**Actual result:** 
photo_2024-05-14 13 33 32


**Expected result:** 

Wallet for Mobile – Figma 2024-05-14 13-32-27

**Comment by pavloburykh** [comment]
@OmarBasem btw, integration tests are still failing.

**Comment by pavloburykh** [comment]
### ISSUE 5 Some transactions are doubled/trippled in Activity tab

**Steps:** 

1. Perform transaction 
2. Check transaction in Activity tab (pay attention if it is not doubled/tripled)

**Actual result:** transaction is tripled

photo_2024-05-14 15 20 28

**Expected result:** there is only one outgoing transaction at this time

https://arbiscan.io/address/0x527d22094166ad33e7523d2cbca287798e149c7f

Address 0x527d22094166ad33e7523d2cbca287798e149c7f | Arbiscan 2024-05-14 15-21-26

**Comment by pavloburykh** [comment]
### ISSUE 6 Wrong chain is displayed for transaction in Activity tab

**Steps:** 

1. Perform transaction using Arbitrum network
2. See which network is displayed in Activity tab

**Actual result:** Mainnet is displayed instead of Arbitrum

photo_2024-05-14 15 28 13

**Expected result:** this is Arbotrum transaction, so Arbitrum icon should be shown

https://arbiscan.io/tx/0xe8b75a8a23ab1237792cd29aaaed598a47d271b5057afd4f91247b03e0fb3d9e

**Comment by pavloburykh** [comment]
### ISSUE 7 Type 'Send' is shown fro recipient account when transferring funds between own accounts

Steps: 

1. Transfer funds between own accounts i.e. from Account A to Account B
2. Check incoming transaction in Account B
3. Pay attention at the type of transaction

**Actual result:** the type of transaction os 'Send' though it should be 'Received' 


https://github.com/status-im/status-mobile/assets/97245802/605a0789-fc3d-4e28-94f8-e697969cd1e4


**Expected result:** 
For Account 2 transaction is incoming, so it should be 'Received' instead of 'Send'
Account 1 - outgoing transaction https://arbiscan.io/address/0x49bb53433278b8ff318a04ec2c30cf2e9655f940
Account 2 - incoming transaction https://arbiscan.io/tx/0xe8b75a8a23ab1237792cd29aaaed598a47d271b5057afd4f91247b03e0fb3d9e

**Comment by pavloburykh** [comment]
### ISSUE 8 Transaction shows Pending status in Activity tab despite it is completed

**Steps:** 

1. Perform transaction
2. Wait until Success status in blockchain
3. Check transaction status in Activity tab

**Actual result:** transaction is in Pending

photo_2024-05-14 15 42 48


**Expected result:** transaction should be Completed

https://arbiscan.io/tx/0xe8b75a8a23ab1237792cd29aaaed598a47d271b5057afd4f91247b03e0fb3d9e

**Comment by OmarBasem** [comment]
Thanks for your testing @pavloburykh !

Issue 3 is out of scope. There is an open issue for it: https://github.com/status-im/status-mobile/issues/19853

> @OmarBasem btw, integration tests are still failing.

That's linting 👍 

I will take a look at the other issues

**Comment by pavloburykh** [comment]
@OmarBasem for some reason go version is still outdated in your PR despite you have rebased it. 
I see version v0.179.13 in this PR, but it should be v0.179.16 as in develop. 

Could you please rebase once again so go version be up to date? Thanks.

**Comment by OmarBasem** [comment]
Hey @pavloburykh, 

I have reviewed and fixed the issues.

Issue 4: that icon means finalized. You can check the designs [here](https://www.figma.com/design/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?node-id=22134-223442&t=G5iCpe2GHpiAMAQ9-0)

Issue 5: Please check if it is fixed now

Issue 6: I am displaying the chain as it comes from the backed. It is probably a status-go issue. Please note that the current method we are using for fetching activities is deprecated and will be replaced soon https://github.com/status-im/status-mobile/issues/19864

Issue 7: Fixed!

Issue 8: For me it works as expected. Please note that currently we fetch/refresh the activities when a user presses on the account card.

**Comment by pavloburykh** [comment]
@OmarBasem thanks for the PR. Issues 2, 5, 7 seem to be fixed. It is hard to test functionality as balances are barely fetched, activities sometimes fetched, sometimes not. I do not want to block the PR anymore, so let's merge it. Will continue more thorough testing once https://github.com/status-im/status-mobile/issues/19864 and https://github.com/status-im/status-mobile/issues/20011 are fixed.

**Pull request merged**
